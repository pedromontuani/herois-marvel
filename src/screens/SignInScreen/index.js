import React, { useState } from 'react';
import {
  Text,
  View,
  TextInput,
  ImageBackground,
  Image,
  ActivityIndicator,
} from 'react-native';

import { useSelector, useDispatch } from 'react-redux';
import { SafeAreaView } from 'react-native-safe-area-context';
import authSelector from '~/store/modules/auth/selectors';
import { login } from '~/store/modules/auth/actions';

import RoundButton from '~/components/RoundButton';

import colors from '~/theme/colors';
import styles from './styles';

const SignInScreen = () => {
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [emailValid, setEmailValid] = useState(true);
  const [passwordValid, setPasswordValid] = useState(true);
  const isLoading = useSelector(authSelector.isLoading);

  const dispatch = useDispatch();

  const onChangeEmail = (text) => {
    setEmail(text);
    setEmailValid(true);
  };

  const onChangePassword = (text) => {
    setPassword(text);
    setPasswordValid(true);
  };

  const validate = () => {
    const emailValidation = email.match(/^[^\s@]+@[^\s@]+$/);
    const passwordValidation = !!password.trim();
    setEmailValid(emailValidation);
    setPasswordValid(passwordValidation);

    return emailValidation && passwordValidation;
  };

  const onSubmit = () => {
    if (validate()) {
      dispatch(login(email, password));
    }
  };

  return (
    <ImageBackground
      style={styles.viewport}
      source={require('~/assets/login-background/background.png')}
      resizeMode="cover"
    >
      <SafeAreaView style={styles.container}>
        <View style={styles.logoHolder}>
          <Image
            style={styles.firebaseLogo}
            source={require('~/assets/logo_ioasys.png')}
            resizeMode="contain"
          />
        </View>

        <View style={styles.inputsHolder}>
          <TextInput
            style={styles.textInput}
            underlineColorAndroid={emailValid ? colors.white : colors.error}
            placeholderTextColor={emailValid ? colors.white : colors.error}
            placeholder="Email"
            autoCompleteType="email"
            keyboardType="email-address"
            textContentType="emailAddress"
            autoCapitalize="none"
            onChangeText={onChangeEmail}
          />

          <TextInput
            style={styles.textInput}
            underlineColorAndroid={passwordValid ? colors.white : colors.error}
            placeholderTextColor={passwordValid ? colors.white : colors.error}
            placeholder="Senha"
            textContentType="password"
            secureTextEntry
            onChangeText={onChangePassword}
          />
        </View>

        <View style={styles.buttonsHolder}>
          <RoundButton
            backgroundColor={colors.white}
            borderColor={colors.transparent}
            disabled={isLoading}
            onPress={onSubmit}
          >
            {isLoading ? (
              <ActivityIndicator size="large" color={colors.primary} />
            ) : (
              <Text style={styles.textFullButton}>LOGIN</Text>
            )}
          </RoundButton>
        </View>
      </SafeAreaView>
    </ImageBackground>
  );
};

export default SignInScreen;
