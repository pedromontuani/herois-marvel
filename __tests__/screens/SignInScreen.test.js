import React from 'react';
import renderer from 'react-test-renderer';
import { fireEvent, render, waitFor } from '@testing-library/react-native';

import {
  rightPassword,
  rigthEmail,
  wrongEmails,
  wrongPassword
} from '../__mocks__/screens/SignInScreen.mock';
import { mockDispatch } from '../__mocks__/libs/ReactRedux.mock';
import SignInScreen from '~/screens/SignIn';
import providersHoc from '../utils/providersHoc';

describe('SignInScreen', () => {
  it('should render correctly', () => {
    const screen = providersHoc(<SignInScreen />);
    const rendered = renderer.create(screen).toJSON();
    expect(rendered).toMatchSnapshot();
  });

  it('should not submit with wrong email and right password', async () => {
    const screen = providersHoc(<SignInScreen />);
    const { getByTestId } = render(screen);

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitButton = getByTestId('submit-button');

    fireEvent.changeText(passwordInput, rightPassword);
    for await (const key of Object.keys(wrongEmails)) {
      fireEvent.changeText(emailInput, wrongEmails[key]);
      fireEvent.press(submitButton);
      await waitFor(() => expect(mockDispatch).not.toHaveBeenCalled());
    }
  });

  it('should not submit with right email and wrong password', async () => {
    const screen = providersHoc(<SignInScreen />);
    const { getByTestId } = render(screen);

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitButton = getByTestId('submit-button');

    fireEvent.changeText(emailInput, rigthEmail);
    fireEvent.changeText(passwordInput, wrongPassword);
    fireEvent.press(submitButton);
    await waitFor(() => expect(mockDispatch).not.toHaveBeenCalled());
  });

  it('should submit with right email and password', async () => {
    const screen = providersHoc(<SignInScreen />);
    const { getByTestId } = render(screen);

    const emailInput = getByTestId('email-input');
    const passwordInput = getByTestId('password-input');
    const submitButton = getByTestId('submit-button');

    fireEvent.changeText(emailInput, rigthEmail);
    fireEvent.changeText(passwordInput, rightPassword);
    fireEvent.press(submitButton);
    await waitFor(() => expect(mockDispatch).toHaveBeenCalled());
  });
});
