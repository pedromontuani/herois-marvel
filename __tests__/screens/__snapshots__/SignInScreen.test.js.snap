// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInScreen should render correctly 1`] = `
<View
  accessibilityIgnoresInvertColors={true}
  style={
    Array [
      Object {
        "height": 1334,
      },
    ]
  }
>
  <Image
    resizeMode="cover"
    source={
      Object {
        "testUri": "../../../src/assets/login-background/background.png",
      }
    }
    style={
      Array [
        Object {
          "bottom": 0,
          "left": 0,
          "position": "absolute",
          "right": 0,
          "top": 0,
        },
        Object {
          "height": undefined,
          "width": undefined,
        },
        undefined,
      ]
    }
  />
  <RNCSafeAreaView
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexBasis": 0,
          "flexDirection": "column",
          "flexGrow": 1,
          "flexShrink": 1,
          "justifyContent": "space-between",
          "paddingBottom": 30,
          "paddingLeft": 30,
          "paddingRight": 30,
          "paddingTop": 30,
        },
      ]
    }
  >
    <View
      style={
        Array [
          Object {
            "flexBasis": 0,
            "flexGrow": 0.3,
            "flexShrink": 1,
            "paddingTop": 20,
          },
        ]
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/assets/logo_ioasys.png",
          }
        }
        style={
          Array [
            Object {
              "height": "60%",
            },
          ]
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 0.3,
            "flexShrink": 1,
            "justifyContent": "space-around",
          },
        ]
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCompleteType="email"
        keyboardType="email-address"
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Email"
        placeholderTextColor="#ffffff"
        rejectResponderTermination={true}
        style={
          Array [
            Object {
              "color": "#ffffff",
              "paddingBottom": 20,
              "width": "100%",
            },
            Object {
              "borderBottomColor": "#ffffff",
              "borderBottomWidth": 1,
            },
            undefined,
          ]
        }
        testID="email-input"
        textContentType="emailAddress"
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        onBlur={[Function]}
        onChangeText={[Function]}
        placeholder="Senha"
        placeholderTextColor="#ffffff"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Array [
            Object {
              "color": "#ffffff",
              "paddingBottom": 20,
              "width": "100%",
            },
            Object {
              "borderBottomColor": "#ffffff",
              "borderBottomWidth": 1,
            },
            undefined,
          ]
        }
        testID="password-input"
        textContentType="password"
        underlineColorAndroid="transparent"
        value=""
      />
    </View>
    <View
      style={
        Array [
          Object {
            "alignSelf": "stretch",
            "flexBasis": 0,
            "flexDirection": "column",
            "flexGrow": 0.4,
            "flexShrink": 1,
            "justifyContent": "center",
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "height": 131.4,
              "width": "100%",
            },
          ]
        }
      >
        <View
          accessible={true}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="submit-button"
        >
          <View
            style={
              Array [
                Object {
                  "borderRadius": 118.26,
                  "borderStyle": "solid",
                  "borderWidth": 2,
                  "height": "90%",
                },
                Object {
                  "backgroundColor": "#ffffff",
                  "borderColor": "transparent",
                },
              ]
            }
          >
            <View
              style={
                Array [
                  Object {
                    "alignItems": "center",
                    "flexBasis": 0,
                    "flexDirection": "column",
                    "flexGrow": 1,
                    "flexShrink": 1,
                    "justifyContent": "space-around",
                  },
                ]
              }
            >
              <View
                style={
                  Array [
                    Object {
                      "alignSelf": "stretch",
                      "flexDirection": "row",
                      "justifyContent": "center",
                    },
                  ]
                }
              >
                <Text
                  style={
                    Array [
                      Object {
                        "color": "#5F49AB",
                        "fontWeight": "bold",
                      },
                    ]
                  }
                >
                  LOGIN
                </Text>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </RNCSafeAreaView>
</View>
`;
